File: report.py
"""Write match results to a color-coded Excel report."""
from __future__ import annotations

from pathlib import Path

import openpyxl
from openpyxl.styles import Font, PatternFill

from models import MatchResult

STATUS_TEXT = {
    "OK": "OK",
    "WARNING": "rozdíl",
    "MISSING": "chybí na faktuře",
    "EXTRA": "navíc na faktuře",
}
_FILLS = {
    "OK": PatternFill("solid", start_color="C6EFCE"),       # green
    "WARNING": PatternFill("solid", start_color="FFEB9C"),  # orange
    "MISSING": PatternFill("solid", start_color="FFC7CE"),  # red
    "EXTRA": PatternFill("solid", start_color="FFC7CE"),    # red
}

HEADERS = ["Stav", "Objekt", "Pozice", "Rozměr objednávka", "Rozměr faktura",
           "Ks objednáno", "Ks fakturováno", "Skladba objednávka",
           "Skladba faktura", "Problémy"]


def result_row(r: MatchResult) -> list[str]:
    o, i = r.order_item, r.invoice_item
    return [
        STATUS_TEXT[r.status],
        o.objekt if o else "",
        o.label if o else (i.label if i else ""),
        f"{o.width} x {o.height}" if o else "",
        f"{i.width} x {i.height}" if i else "",
        str(o.quantity) if o else "",
        str(i.quantity) if i else "",
        o.skladba_raw if o else "",
        f"{i.composition_raw} | rámeček {i.spacer} mm" if i else "",
        "; ".join(r.problems),
    ]


def sheet_title(results: list[MatchResult]) -> str:
    """Which supplier this report came from. Excel caps sheet names at 31."""
    provider = next((r.invoice_item.provider for r in results if r.invoice_item), "")
    return f"Kontrola — {provider}"[:31] if provider else "Kontrola"


def write_report(results: list[MatchResult], path: str | Path) -> None:
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = sheet_title(results)
    ws.append(HEADERS)
    for cell in ws[1]:
        cell.font = Font(bold=True)
    for r in results:
        ws.append(result_row(r))
        fill = _FILLS[r.status]
        for cell in ws[ws.max_row]:
            cell.fill = fill
    for col, width in zip("ABCDEFGHIJ", (16, 16, 10, 16, 16, 12, 12, 24, 40, 50)):
        ws.column_dimensions[col].width = width
    wb.save(path)

File: web/lib/report.js
// Port report.py. resultRow slouží tabulce na obrazovce i sešitu, stejně
// jako result_row v Pythonu.

export const STATUS_TEXT = {
  OK: "OK",
  WARNING: "rozdíl",
  MISSING: "chybí na faktuře",
  EXTRA: "navíc na faktuře",
};

const VYPLNE = {
  OK: "FFC6EFCE",        // zelená
  WARNING: "FFFFEB9C",   // oranžová
  MISSING: "FFFFC7CE",   // červená
  EXTRA: "FFFFC7CE",
};

export const HEADERS = ["Stav", "Objekt", "Pozice", "Rozměr objednávka",
  "Rozměr faktura", "Ks objednáno", "Ks fakturováno", "Skladba objednávka",
  "Skladba faktura", "Problémy"];

const SIRKY = [16, 16, 10, 16, 16, 12, 12, 24, 40, 50];

export function resultRow(r) {
  const o = r.orderItem;
  const i = r.invoiceItem;
  return [
    STATUS_TEXT[r.status],
    o ? o.objekt : "",
    o ? o.label : (i ? i.label : ""),
    o ? `${o.width} x ${o.height}` : "",
    i ? `${i.width} x ${i.height}` : "",
    o ? String(o.quantity) : "",
    i ? String(i.quantity) : "",
    o ? o.skladbaRaw : "",
    i ? `${i.compositionRaw} | rámeček ${i.spacer} mm` : "",
    r.problems.join("; "),
  ];
}

/** ExcelJS se předává zvenčí, stejně jako u order-parseru. */
/** Od kterého dodavatele report je. Excel má strop 31 znaků na název listu. */
export function sheetTitle(results) {
  const r = results.find((v) => v.invoiceItem);
  const provider = r ? r.invoiceItem.provider : "";
  return provider ? `Kontrola — ${provider}`.slice(0, 31) : "Kontrola";
}

export async function reportBuffer(ExcelJS, results) {
  const wb = new ExcelJS.Workbook();
  const ws = wb.addWorksheet(sheetTitle(results));
  ws.addRow(HEADERS).font = { bold: true };
  for (const r of results) {
    const row = ws.addRow(resultRow(r));
    row.eachCell({ includeEmpty: true }, (cell) => {
      cell.fill = { type: "pattern", pattern: "solid",
                    fgColor: { argb: VYPLNE[r.status] } };
    });
  }
  SIRKY.forEach((sirka, i) => { ws.getColumn(i + 1).width = sirka; });
  return wb.xlsx.writeBuffer();
}
